From 5f4fc1c04c95d8ac34fcd094cfe23cb35cd43290 Mon Sep 17 00:00:00 2001
From: Pawit Pornkitprasan <p.pawit@gmail.com>
Date: Fri, 28 Nov 2014 18:11:52 +0700
Subject: [PATCH] Blacklist: fix NPE

getCallerInfo() can be null if the caller info is being
asynchronously read. Use a different method to get the phone number.

Change-Id: I4b77552a3e3cddd6d3fc9ba8e13721d7431f6141
---
 src/com/android/server/telecom/Call.java         | 4 ++++
 src/com/android/server/telecom/CallsManager.java | 7 +++++--
 2 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/src/com/android/server/telecom/Call.java b/src/com/android/server/telecom/Call.java
index eeb592e..79fed09 100644
--- a/src/com/android/server/telecom/Call.java
+++ b/src/com/android/server/telecom/Call.java
@@ -1217,6 +1217,10 @@ final class Call implements CreateConnectionResponse {
         return mCallerInfo;
     }
 
+    String getNumber() {
+        return mHandle == null ? null : mHandle.getSchemeSpecificPart();
+    }
+
     /**
      * Saves the specified photo information if the specified token matches that of the last query.
      *
diff --git a/src/com/android/server/telecom/CallsManager.java b/src/com/android/server/telecom/CallsManager.java
index 33b91c7..ef61d0e 100644
--- a/src/com/android/server/telecom/CallsManager.java
+++ b/src/com/android/server/telecom/CallsManager.java
@@ -39,7 +39,6 @@ import android.telecom.TelecomManager;
 import android.telephony.TelephonyManager;
 
 import com.android.internal.telephony.CallStateException;
-import com.android.internal.telephony.CallerInfo;
 import com.android.internal.telephony.Connection;
 import com.android.internal.telephony.TelephonyProperties;
 import com.android.internal.telephony.util.BlacklistUtils;
@@ -1829,7 +1828,11 @@ public final class CallsManager extends Call.ListenerBase {
     }
 
     protected boolean isCallBlacklisted(Call c) {
-        final String number = c.getCallerInfo().phoneNumber;
+        final String number = c.getNumber();
+        if (number == null) {
+            return false;
+        }
+
         // See if the number is in the blacklist
         // Result is one of: MATCH_NONE, MATCH_LIST or MATCH_REGEX
         int listType = BlacklistUtils.isListed(mContext, number, BlacklistUtils.BLOCK_CALLS);
-- 
1.9.3 (Apple Git-50)

